[Lens] Implement basic editor frame state handling

flash1293 opened the pull request:
Based on https://github.com/elastic/kibana/pull/36165 (has to be merged first)

Related to https://github.com/elastic/kibana/issues/36184

This sets up basic state handling:
* initializing datasource and visualization
* making switching possible (for debugging)
* making state updates possible
* passing the datasource api to the visualization

Suggestion handling is missing, but I will address them in a separate PR because this one already got a bit large.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13555/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13567/)

flash1293 commented:
This also contains the changes for the native renderer itself, it will go away once the other PR is merged. I will clean that up next week, but you can already check out the structure and implementation inside `editor_frame_plugin`

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13584/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13730/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13734/)

flash1293 commented:
@wylieconlon @chrisdavies 

I find myself passing down a lot of stuff from the frame component itself which holds the state to sub components which I introduced to split code up a little:
```
      <FrameLayout
        dataPanel={
          <DataPanelWrapper
            activeDatasource={state.activeDatasource}
            datasourceIsLoading={state.datasourceIsLoading}
            datasourceState={state.datasourceState}
            datasources={props.datasources}
            dispatch={dispatch}
          />
        }
        configPanel={
          <ConfigPanelWrapper
            visualizations={props.visualizations}
            activeVisualization={state.activeVisualization}
            datasourcePublicAPI={datasourcePublicAPI!}
            dispatch={dispatch}
            visualizationState={state.visualizationState}
          />
        }
        suggestionsPanel={<SuggestionPanelWrapper suggestions={suggestions} dispatch={dispatch} />}
      />
```

An alternative I'm thinking about is introducing a context and basically doing something like redux. I'm not sure whether this is necessary here or not. The editor frame WILL grow in complexity, but probably not too much, since most of the logic will live in the datasource and the visualizations. Currently there are two layers - the editor frame itself and a config/data/suggestion/workspace panel wrapper. We probably want to add one more layer below that when adding all the bells and whistles. Basically my question is "is props drilling fine for the limited scope of the editor frame with 2-3 layers of components or does a context make more sense here?"

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13969/)

chrisdavies commented on x-pack/plugins/lens/public/editor_frame_plugin/state_management.ts:
```diff
@@ -0,0 +1,98 @@
+        datasourceState: action.newState,
+      };
+    case 'UPDATE_VISUALIZATION_STATE':
+      if (state.activeVisualization) {
```
Bit of a nitpick, but I find this kind of logic easier to follow if validations are first, and exceptional cases exit early, rather than having them at the end:

```js
if (!state.activeVisualization) {
  throw new Error('Invariant: visualization state got updated without active visualization');
}

return {
  ...state,
  visualizationState: {
    ...state.visualizationState,
    [state.activeVisualization]: action.newState,
  },
};
```

wylieconlon commented on x-pack/plugins/lens/public/editor_frame_plugin/editor_frame/editor_frame.test.tsx:
```diff
@@ -0,0 +1,441 @@
+import { Visualization, Datasource, DatasourcePublicAPI } from '../../types';
+import { act } from 'react-dom/test-utils';
+
+const nextTick = () => new Promise(resolve => setTimeout(resolve));
```
Could you use timer mocks instead of having this? https://jestjs.io/docs/en/timer-mocks

flash1293 replied:
No, because the purpose of this is to wait for the Promise result of the mocked datasource (`    initialize: jest.fn(() => Promise.resolve()),`) which is asynchronous. The timer mocks only monkey patch the timer functions, not Promises. We could mock Promises by ourselves to behave synchronously, but I would rather not hide the asynchronous nature of Promises in the test as it might shadow certain conditions. I changed the implementation to `const waitForPromises = () => new Promise(resolve => setImmediate(resolve));` which does the same thing and added a comment.

wylieconlon commented:
Regarding your earlier question on props drilling: yes, I think it's a little more explicit for now. Might have issues soon, but we can change that part.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14228/)

flash1293 commented on x-pack/plugins/lens/public/editor_frame_plugin/state_management.ts:
```diff
@@ -0,0 +1,98 @@
+
+export const getInitialState = (props: EditorFrameProps) => {
+  return {
+    datasourceState: null,
```
We basically store multiple visualization states to be able to switch back and forth between them without loosing any configuration settings which are not related to the table columns.
Example: You have an xy chart with stack bars and switch to a pie chart. Now you switch back to the xy chart, but the settings of using stacked bars has been deleted and you are back to unstacked lines.

We do this for the visualizations because it's really common to switch around between them, but it's not common to switch between datasources after you started configuring. In fact this might not even really make sense, because they are just so different - what should happen when we switch from index pattern datasource to csv datasource? 

Working with two top level props looks nice though, I changed the shape to
```
export interface EditorFrameState {
  visualization: {
    activeId: string | null;
    stateMap: {
      [visualizationId: string]: unknown;
    };
  };
  datasource: {
    activeId: string | null;
    state: unknown;
    isLoading: boolean;
  };
}
```

I also renamed all props which contain a map of things (`Record<string, T>`) to ...Map to make the relationship clear.

flash1293 commented on x-pack/plugins/lens/public/editor_frame_plugin/editor_frame/config_panel_wrapper.tsx:
```diff
@@ -0,0 +1,63 @@
+      />
+      {props.activeVisualization && (
+        <NativeRenderer
+          render={props.visualizations[props.activeVisualization].renderConfigPanel}
```
But it is only used once and if we do this at the root we have to deal with the typescript error because `props.activeVisualization` can be null

flash1293 commented on x-pack/plugins/lens/public/editor_frame_plugin/editor_frame/editor_frame.tsx:
```diff
@@ -0,0 +1,88 @@
+
+  const datasourcePublicAPI = useMemo(
+    () =>
+      state.activeDatasource && !state.datasourceIsLoading
```
Makes sense, I added the comment and changed the shape of the internal state

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14255/)

flash1293 commented:
Jenkins, test this please. Seems like unrelated error

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14275/)

flash1293 commented:
@wylieconlon @chrisdavies I updated the PR to just handle a single visualization state. As I changed quite a few things since the last review, please read through it again if you have the time.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14316/)
